### PR TITLE
feat: persistent configuration to disk (FF-2243)

### PR DESCRIPTION
**Pull request by leoromanovsky**
## Motivation

* When the SDK is initialized, it might not have network access. This is particularly true on mobile. Users of the SDK want to the last downloaded configuration to be cached on disk so they can provide the same UX until network becomes available again.

## Changes

Augment the `setConfiguration` method to update disk as well as memory.

Writing to disk is a fairly low-level interface but straight-forward since we are treating the configuration as a blob and not interested in loading from it, so a database is not needed.

The synchronization on the method is [DispatchQueue.asyncAndWait](https://developer.apple.com/documentation/dispatch/dispatchqueue/3191901-asyncandwait), which is blocking but is called from an async method:

```
    // Loads the configuration from the remote source on-demand. Can be used to refresh as desired.
    //
    // This function can be called from multiple threads; synchronization is provided to safely update
    // the configuration cache but each invocation will execute a new network request with billing impact.
    public func load() async throws {
        let config = try await self.configurationRequester.fetchConfigurations()
        self.configurationStore.setConfiguration(configuration: config)
    }
```

Configuration is loaded from disk on class init, which will slow down `EppoClient` initialize. In the case of `initialConfiguration` provided there will be an additional write as well.

No changes are made to the read (assignment path), this continues to be served only from memory.

### CODE REVIEW

File: Sources/eppo/Configuration.swift [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import Foundation
 
-public struct Configuration {
+public struct Configuration: Codable {
```
Comment (leoromanovsky) [inline Sources/eppo/Configuration.swift:3]: `Codable` interface is for converting objects to and from json.

File: Sources/eppo/ConfigurationStore.swift [hunk 2]
```diff
@@ -2,11 +2,20 @@ import Foundation
...
 class ConfigurationStore {
     private var configuration: Configuration?
-    
     private let syncQueue = DispatchQueue(
         label: "com.eppo.configurationStoreQueue", attributes: .concurrent)
     
+    private let fileURL: URL
+    
+    // Initialize with the disk-based path for storage
     public init() {
```
Comment (typotter) [inline Sources/eppo/ConfigurationStore.swift:11]: What do you think about passing through the `initialConfiguration` optional parameter available from [`EppoClient`](https://github.com/Eppo-exp/eppo-ios-sdk/pull/42/files#diff-dc60401cf877df4122e1cb0b73c011aba7a9a665785dc943c9855d104de4d7e9R69)? We can skip the initial disk read when there's a configuration string.

File: Sources/eppo/dto/EppoValue.swift [hunk 3]
```diff
@@ -188,3 +188,20 @@ public class EppoValue : Decodable, Equatable {
         }
     }
 }
+extension EppoValue {
```
Comment (typotter) [inline Sources/eppo/dto/EppoValue.swift:191]: Is this to support the `Codable` interface?

**Review by typotter — APPROVED** [review]
Looking good! Just one suggestion to consider.

File: Sources/eppo/dto/EppoValue.swift (hunk 3 above)
Comment (leoromanovsky) [reply Sources/eppo/dto/EppoValue.swift:191]: Yes

File: Sources/eppo/ConfigurationStore.swift (hunk 2 above)
Comment (leoromanovsky) [reply Sources/eppo/ConfigurationStore.swift:11]: Good idea!

**Review by aarsilv — APPROVED** [review]
Nice job getting this in too!

File: Tests/eppo/ConfigurationStoreTests.swift [hunk 4]
```diff
@@ -30,6 +33,24 @@ final class ConfigurationStoreTests: XCTestCase {
...
     
+    override func tearDownWithError() throws {
+        try super.tearDownWithError()
+        // Clean up the configuration file
+        cleanUpConfigurationFile()
+        // Release the reference to ensure it's deallocated
+        configurationStore = nil
+    }
+    
+    private func cleanUpConfigurationFile() {
```
Comment (rasendubi) [inline Tests/eppo/ConfigurationStoreTests.swift:44]: nit: does it make sense to add this to `ConfigurationStore`? so the test doesn't need to know where and how configuration is stored

File: Tests/eppo/ConfigurationStoreTests.swift [hunk 5]
```diff
@@ -30,6 +33,24 @@ final class ConfigurationStoreTests: XCTestCase {
...
+        let fileManager = FileManager.default
+        let urls = fileManager.urls(for: .documentDirectory, in: .userDomainMask)
+        let fileURL = urls[0].appendingPathComponent("configuration.json")
+        
+        if fileManager.fileExists(atPath: fileURL.path) {
+            try? fileManager.removeItem(at: fileURL)
+        }
```
Comment (rasendubi) [inline Tests/eppo/ConfigurationStoreTests.swift:51]: nit: no need to check for existence if we're going to delete the file anyway
```suggestion
        try? fileManager.removeItem(at: fileURL)
```

File: Sources/eppo/ConfigurationStore.swift [hunk 6]
```diff
@@ -2,11 +2,20 @@ import Foundation
...
+        // Set the file path for storing the configuration
+        let fileManager = FileManager.default
+        let urls = fileManager.urls(for: .documentDirectory, in: .userDomainMask)
+        self.fileURL = urls[0].appendingPathComponent("configuration.json")
+        
+        // Load any existing configuration from disk when initializing
+        self.configuration = loadFromDisk()
```
Comment (rasendubi) [inline Sources/eppo/ConfigurationStore.swift:18]: minor: would it make sense to spawn this concurrently, so the store is returned immediately and configuration is populated shortly after?

File: Sources/eppo/ConfigurationStore.swift [hunk 7]
```diff
@@ -2,11 +2,20 @@ import Foundation
 
 class ConfigurationStore {
     private var configuration: Configuration?
-    
     private let syncQueue = DispatchQueue(
         label: "com.eppo.configurationStoreQueue", attributes: .concurrent)
     
+    private let fileURL: URL
```
Comment (rasendubi) [inline Sources/eppo/ConfigurationStore.swift:8]: minor: Adding persistence to `ConfigurationStore` seems obvious. I'm not sure if that's the best place though. One concern I have is that it adds another responsibility to the store and couples persistence with managing global configuration (so they would be harder to separate in the future). e.g., one place where it shows up is complicating tests with the cleanup step

Questions to think about:
- is configuration caching something that users would want to ever disable?
- is it something that users would want to configure? could there be different sensible strategies?

But I'm probably over-complicating this, so feel free to ignore

**Review by rasendubi — APPROVED** [review]

File: Sources/eppo/ConfigurationStore.swift [hunk 8]
```diff
@@ -25,6 +34,30 @@ class ConfigurationStore {
...
+    }
+    
+    // Save the configuration to disk
+    private func saveToDisk(configuration: Configuration) {
+        do {
+            let data = try JSONEncoder().encode(configuration)
+            try data.write(to: fileURL)
+        } catch {
+            print("Error saving configuration to disk: \(error)")
+        }
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply Sources/eppo/ConfigurationStore.swift]: Thanks - I will research this.

File: Sources/eppo/ConfigurationStore.swift (hunk 7 above)
Comment (leoromanovsky) [reply Sources/eppo/ConfigurationStore.swift:10]: It's a good thought for sure. There are some downsides, particularly the disk performance latency. Let me consider making this a default and allowing users to disable it.

Comment (rasendubi) [reply Sources/eppo/ConfigurationStore.swift:10]: If we factor configuration store into interface, the disk could be an adapter/wrapper (similar to caching assignment logger in python)

**Pull request merged**